From ce4c9459b62be487776f4e3b173d02b7486696d5 Mon Sep 17 00:00:00 2001
From: Linu Cherian <lcherian@marvell.com>
Date: Mon, 13 Aug 2018 22:42:59 +0530
Subject: [PATCH 0124/1051] net: octeontx2: Add support to configure loopback

Add support to set/clear CGX in loopback mode using
netdev features.

Signed-off-by: Linu Cherian <linu.cherian@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../net/ethernet/marvell/octeontx2/otx2_pf.c  | 29 +++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/otx2_pf.c b/drivers/net/ethernet/marvell/octeontx2/otx2_pf.c
index 2b5705d74984..47bdf32c7dee 100644
--- a/drivers/net/ethernet/marvell/octeontx2/otx2_pf.c
+++ b/drivers/net/ethernet/marvell/octeontx2/otx2_pf.c
@@ -239,6 +239,21 @@ static int otx2_pfaf_mbox_init(struct otx2_nic *pf)
 	return err;
 }
 
+static int otx2_cgx_config_loopback(struct otx2_nic *pf, bool enable)
+{
+	struct msg_req *msg;
+
+	if (enable)
+		msg = otx2_mbox_alloc_msg_CGX_INTLBK_ENABLE(&pf->mbox);
+	else
+		msg = otx2_mbox_alloc_msg_CGX_INTLBK_DISABLE(&pf->mbox);
+
+	if (!msg)
+		return -ENOMEM;
+
+	return otx2_sync_mbox_msg(&pf->mbox);
+}
+
 static int otx2_set_real_num_queues(struct net_device *netdev,
 				    int tx_queues, int rx_queues)
 {
@@ -643,6 +658,18 @@ static void otx2_reset_task(struct work_struct *work)
 	netif_trans_update(pf->netdev);
 }
 
+static int otx2_set_features(struct net_device *netdev,
+			     netdev_features_t features)
+{
+	struct otx2_nic *pf = netdev_priv(netdev);
+	netdev_features_t changed = features ^ netdev->features;
+
+	if ((changed & NETIF_F_LOOPBACK) && netif_running(netdev))
+		return otx2_cgx_config_loopback(pf,
+						features & NETIF_F_LOOPBACK);
+	return 0;
+}
+
 static const struct net_device_ops otx2_netdev_ops = {
 	.ndo_open		= otx2_open,
 	.ndo_stop		= otx2_stop,
@@ -651,6 +678,7 @@ static const struct net_device_ops otx2_netdev_ops = {
 	.ndo_change_mtu         = otx2_change_mtu,
 	.ndo_set_rx_mode        = otx2_set_rx_mode,
 	.ndo_get_stats64	= otx2_get_stats64,
+	.ndo_set_features	= otx2_set_features,
 	.ndo_tx_timeout         = otx2_tx_timeout,
 };
 
@@ -758,6 +786,7 @@ static int otx2_probe(struct pci_dev *pdev, const struct pci_device_id *id)
 			       NETIF_F_IPV6_CSUM | NETIF_F_RXHASH |
 			       NETIF_F_TSO | NETIF_F_TSO6);
 	netdev->features |= netdev->hw_features;
+	netdev->hw_features |= NETIF_F_LOOPBACK;
 
 	netdev->gso_max_segs = OTX2_MAX_GSO_SEGS;
 
-- 
2.17.1

